gflownet/envs/base.py

```diff
@@ -868,7 +911,7 @@ def set_state(self, state: List, done: Optional[bool] = False):
         at all states (intermediate states are not fully constructed objects) should
         overwrite this method and check for validity.
         """
-        self.state = state
+        self.state = copy(state)
```

Comment:
1. Does this actually need to be copied? 2. If so, this won't work, since some envs (Tree) have tensor states.

Reply:
ABSOLUTELY. This cost me about 4 hours yesterday from 8 pm to midnight :sob:

Reply:
It does work with tensors because this `copy()` is from [here](https://github.com/alexhernandezgarcia/gflownet/blob/backward-sampling-continuous/gflownet/utils/common.py#L235), which catches tensor cases.

Reply:
> It does work with tensors because this copy() is from [here](https://github.com/alexhernandezgarcia/gflownet/blob/backward-sampling-continuous/gflownet/utils/common.py#L235), which catches tensor cases.

Oh, you are right, my bad!

> ABSOLUTELY. This cost me about 4 hours yesterday from 8 pm to midnight

Oof! That's an issue with environments being messed up due to shared states is they happen to be initialized with the same state from a replay buffer/dataset, right?

Reply:
Exactly!

Reply:
And this only became an issue with the estimation of log probabilities because there you do initalise multiple envs (for multiple trajectories) with the very same state. And believe me it failed in the most weird ways...